### PR TITLE
QQbject superclasses implementation

### LINKED ISSUE #197: New superclass for measure-plugin objects
This new super class will be of the form

```python
class QMeasureObjectABC(qtc.QObject, metaclass=base.QMetaABC):
    pass
```

 and will contain common features such as error reporting and detecting appropriate settings. This will be the ancestor class for:
- `MeasurementABC`
- `ControllerABC`
- `CameraABC`

There may be another subclass which is a device base class that contains an abstract list_devices (e.g., a `DeviceABC`) to be used as ancestor for `ControllerABC `and `CameraABC`.

Possibly, an even higher layer for features that are common to even more objects (e.g., `error_occurred`).

### PR DESCRIPTION
**Pull request by FlorianDoerr**
Basic implementation of the new superclasses for ViPErLEED measurement objects.

Fixes #197

### CODE REVIEW

File: guilib/measure/uimeasurement.py [hunk 1]
```diff
@@ -665,13 +683,14 @@ def __make_device(self, device_cls, device_name, **other_info):
         elif issubclass(device_cls, CameraABC):
             section = "camera_settings"
 
+        device_name = settings_info.more['name'] or settings_info.unique_name
         device.settings[section]['device_name'] = device_name
```
Comment (michele-riva) [inline guilib/measure/uimeasurement.py:687]: Potential `NameError` if `section` is not defined (see conditions above). Perhaps add an `else` where you raise a `ValueError`?

File: guilib/measure/uimeasurement.py [hunk 2]
```diff
@@ -665,13 +683,14 @@ def __make_device(self, device_cls, device_name, **other_info):
...
         new_cfg_path = Path(_cfg_dir) / f"{device.name_clean}.ini"
         if new_cfg_path.exists():
             print(f"{section} config file name conflict! Overwriting existing")  # TODO: ask what to do with the (invalid) file
         with new_cfg_path.open('w', encoding='utf-8') as fproxy:
             device.settings.write(fproxy)
-
+        device.uses_default_settings = False
```
Comment (michele-riva) [inline guilib/measure/uimeasurement.py:693]: I now see why you made `.uses_default_settings` public. Discard my comments in `abc` concerning this.

File: guilib/measure/uimeasurement.py [hunk 3]
```diff
@@ -702,7 +721,7 @@ def __on_bad_pixels_selected(self):
 
     def __on_camera_clicked(self, *_):                                          # TODO: may want to display a busy dialog with "starting camera <name>..."
         cam_name = self.sender().text()
-        cam_cls = self.sender().data()
+        cam_cls, cam_info = self.sender().data()
```
Comment (michele-riva) [inline guilib/measure/uimeasurement.py:724]: You probably don't need to unpack these. See my edit below.

File: guilib/measure/uimeasurement.py [hunk 4]
```diff
@@ -712,7 +731,7 @@ def __on_camera_clicked(self, *_):                                          # TO
                 return
 
         # Not already available. Make a new camera.
-        camera = self.__make_device(cam_cls, cam_name)
+        camera = self.__make_device(cam_cls, cam_info)
```
Comment (michele-riva) [inline guilib/measure/uimeasurement.py:734]:
```suggestion
        camera = self.__make_device(*self.sender().data())
```

**Review by michele-riva — CHANGES_REQUESTED** [review]
Great work @FlorianDoerr. A few small changes, but we're almost there!

**Comment by michele-riva** [comment]
Please also make sure to merge in the base branch and resolve conflicts.

File: guilib/measure/camera/imagingsource.py [hunk 5]
```diff
@@ -529,6 +531,85 @@ def close(self):
         """Close the camera device."""
         self.driver.close()
 
+    @classmethod
+    def is_matching_default_settings(cls, obj_info, config, match_exactly):
+        """Determine if the default settings file is for this camera.
```
[the remark this answers is not in the document]
Comment (FlorianDoerr) [reply guilib/measure/camera/imagingsource.py:539]: I updated the Sphinx related changes but left the rest, as this method only returns (1,) in any case.

File: guilib/measure/classes/abc.py [hunk 6]
```diff
@@ -377,12 +399,12 @@ def are_settings_invalid(self, new_settings):
...
         invalid_settings : list of tuples
-            Invalid required_settings of self as a list of tuples.
+            Invalid _mandatory_settings of self as a list of tuples.
             The first entry in each tuple can be either '<section>',
             '<section>/<option>', or
             '<section>/<option> not one of <value1>, <value2>, ...'.
-            Further entries are information on what is wrong with
-            the setttings.
+            Further entries, added by reimplementations, are
+            information on what is wrong with the setttings.
```
Comment (michele-riva) [inline guilib/measure/classes/abc.py:407]:
```suggestion
            information on what is wrong with the settings.
```

File: guilib/measure/classes/abc.py [hunk 7]
```diff
@@ -0,0 +1,568 @@
...
+        Emits
+        -----
+        QObjectSettingsErrors.MISSING_SETTINGS
+            If the settings have not been found
+            or if new_settings was None.
+        QObjectSettingsErrors.INVALID_SETTINGS
+            If any element of the new_settings does
+            not fit the mandatory settings.
+        """
+        try:                                                                    # TODO: make method that searches through invalid for old values and replaces deprecated ones, make it a method of the ViPErLEEDSettings class
```
[the remark this answers is not in the document]
Comment (michele-riva) [reply guilib/measure/classes/abc.py:468]: I'd suggest to replace the TODO with `# TODO: #242`

File: guilib/measure/controller/abc.py [hunk 8]
```diff
@@ -467,20 +467,20 @@ def time_to_trigger(self):
 
     def _update_serial_from_settings(self):
         """Set serial settings from new controller settings."""
-        serial_cls_name = self._settings.get('controller', 'serial_class',
-                                             fallback='')                       # TODO: remove fallback in 1.0
-        if not serial_cls_name:                                                 # TODO: only here for backwards compatibility, remove in 1.0
-            serial_cls_name = self._settings.get('controller',
+        serial_cls_name = self.settings.get('controller', 'serial_class',
+                                             fallback='')                       # TODO: #<242> (remove fallback in 1.0)
```
Comment (michele-riva) [inline guilib/measure/controller/abc.py:471]: indent

**Review by michele-riva — CHANGES_REQUESTED** [review]
Sorry, I wanted to add single comments, but ended up clicking the wrong button and created another review.

File: guilib/measure/measurement/abc.py [hunk 9]
```diff
@@ -268,17 +253,66 @@ def secondary_controllers(self, new_controllers):
...
+            The settings to check.
+        match_exactly : bool
+            Whether obj_info should be matched exactly.
+
+        Returns
+        -------
+        is_suitable : bool
+            True if the settings file is suitable.
+        """
+        return ()
```
[the remark this answers is not in the document]
Comment (FlorianDoerr) [reply guilib/measure/measurement/abc.py:274]: This just was not used and not implemented for measurements back then. Has been implemented in the measurement-settings-dialog branch.

File: guilib/measure/controller/abc.py [hunk 10]
```diff
@@ -468,10 +468,10 @@ def time_to_trigger(self):
     def _update_serial_from_settings(self):
         """Set serial settings from new controller settings."""
         serial_cls_name = self.settings.get('controller', 'serial_class',
-                                             fallback='')                       # TODO: #<242> (remove fallback in 1.0)
-        if not serial_cls_name:                                                 # TODO: #<242> (only here for backwards compatibility, remove in 1.0)
+                                            fallback='')                       # TODO: #242 (remove fallback in 1.0)
+        if not serial_cls_name:                                                 # TODO: #242 (only here for backwards compatibility, remove in 1.0)
```
Comment (michele-riva) [inline guilib/measure/controller/abc.py:472]: Once we have the aliases in place as in #242, I think we can leave this part in. I would remove the "remove XXX in 1.0"

File: guilib/measure/dialogs/constants.py [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+
+Created: 2024-10-10
+Author: Michele Riva
+Author: Florian Doerr
+
+Defines constants used in the other ViPErLEED modules.
+"""
+
+
+DIALOG_DISMISSED = object()
```
Comment (michele-riva) [inline guilib/measure/dialogs/constants.py:15]: Missing newline at the end

File: guilib/measure/dialogs/constants.py [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+
+========================================
+   ViPErLEED Graphical User Interface
+========================================
+
+Created: 2024-10-10
+Author: Michele Riva
+Author: Florian Doerr
+
+Defines constants used in the other ViPErLEED modules.
```
Comment (michele-riva) [inline guilib/measure/dialogs/constants.py:11]:
```suggestion
Defines dialog-related constants used in the other ViPErLEED modules.
```

File: guilib/measure/dialogs/constants.py [hunk 13]
```diff
@@ -0,0 +1,15 @@
+"""Module constants of viperleed.guilib.measure.dialogs.
```
Comment (michele-riva) [inline guilib/measure/dialogs/constants.py:1]: Consider if it makes sense to move this module "one level up", to guilib/dialogs

File: guilib/measure/hardwarebase.py [hunk 14]
```diff
@@ -314,7 +315,7 @@ def get_object_settings(obj_cls, obj_info, **kwargs):
         The path to the only settings file successfully found or the
         file that has been selected by the user. None if no settings
         file was found. None is returned if the user dismissed the
-        pop-up. The empty string is returned if an alternative option
+        pop-up. DIALOG_DISMISSED is returned if an alternative option
```
Comment (michele-riva) [inline guilib/measure/hardwarebase.py:318]: Add this to the type specification too (I cannot comment on that line):

```python
    path_to_config : Path or None or DIALOG_DISMISSED
```

File: guilib/measure/measurement/abc.py [hunk 15]
```diff
@@ -255,48 +255,58 @@ def secondary_controllers(self, new_controllers):
...
             The settings to check.
         match_exactly : bool
             Whether obj_info should be matched exactly.
 
         Returns
         -------
-        is_suitable : bool
-            True if the settings file is suitable.
+        sorting_info : tuple
+            A tuple that can be used the sort the detected settings.
```
Comment (michele-riva) [inline guilib/measure/measurement/abc.py:272]:
```suggestion
            A tuple that can be used to sort the detected settings.
```

Comment (michele-riva) [reply guilib/measure/measurement/abc.py:272]: Multiple occurrences

File: guilib/measure/measurement/abc.py [hunk 16]
```diff
@@ -1139,6 +1150,9 @@ class to be instantiated.
         # Backwards compatibility fix for port_name:
```
Comment (michele-riva) [inline guilib/measure/measurement/abc.py:1150]: Add a `# TODO: #242` here too

**Review by michele-riva — CHANGES_REQUESTED** [review]
Last round, I promise!

File: guilib/measure/classes/abc.py [hunk 17]
```diff
@@ -0,0 +1,603 @@
...
+        Returns
+        -------
+        is_suitable : bool
+            True if the settings file is for the class.
+        """
+
+    def are_settings_invalid(self, new_settings):
+        """Check if there are any invalid settings.
+
+        Reimplementations can add additional mandatory settings at
```
Comment (michele-riva) [inline guilib/measure/classes/abc.py:393]:
```suggestion
        Subclasses may add additional mandatory settings at
```

Comment (michele-riva) [reply guilib/measure/classes/abc.py:393]: Perhaps an indication on how to do that would be useful. Either here or (better) in the description of the `_mandatory_settings` attribute above.

File: guilib/measure/classes/abc.py [hunk 18]
```diff
@@ -0,0 +1,603 @@
...
+
+        Returns
+        -------
+        invalid_settings : list of tuples
+            Invalid _mandatory_settings of self as a list of tuples.
+            The first entry in each tuple can be either '<section>',
+            '<section>/<option>', or
+            '<section>/<option> not one of <value1>, <value2>, ...'.
+            Further entries, added by reimplementations, are
+            information on what is wrong with the settings.
```
Comment (michele-riva) [inline guilib/measure/classes/abc.py:410]:
```suggestion
            Further optional entries may be added by subclasses. They
            specify additional information on what is wrong with each
            invalid setting.
```

File: guilib/measure/classes/abc.py [hunk 19]
```diff
@@ -0,0 +1,603 @@
...
+        -------
+        handler : SettingsHandler
+            The handler used in a SettingsDialog to display the
+            settings of this instance to users.
+        """
+        self.check_creating_settings_handler_is_possible()
+        handler = SettingsHandler(self.settings)
+        return handler
+
+    def set_settings(self, new_settings):
```
Comment (michele-riva) [inline guilib/measure/classes/abc.py:437]: Do we want to make this a `pyqtSlot`? If yes, make sure to propagate decoration to subclasses that extend this.

File: guilib/measure/classes/abc.py [hunk 20]
```diff
@@ -0,0 +1,603 @@
...
+        Emits
+        -----
+        QObjectSettingsErrors.MISSING_SETTINGS
+            If the settings have not been found
+            or if new_settings was None.
+        QObjectSettingsErrors.INVALID_SETTINGS
+            If any element of the new_settings does
+            not fit the mandatory settings.
+        """
+        try:                                                                    # TODO: make method that searches through invalid for old values and replaces deprecated ones, make it a method of the ViPErLEEDSettings class
```
Comment (michele-riva) [inline guilib/measure/classes/abc.py:471]:
```suggestion
        try:                                                                    # TODO: #242 make method that searches through invalid for old values and replaces deprecated ones, make it a method of the ViPErLEEDSettings class
```

File: guilib/measure/classes/abc.py [hunk 21]
```diff
@@ -0,0 +1,603 @@
...
+            True if the instance is busy.
+
+        Emits
+        -----
+        busy_changed
+            If the busy state of the instance changed.
+        """
+        self.set_busy(is_busy)
+
+    def get_busy(self):
```
Comment (michele-riva) [inline guilib/measure/classes/abc.py:525]: Do we want this to be public to the outside (i.e., not only in subclasses)? Perhaps better to make it a `_get_busy` instead? The only reason to make it public would be to make it also a `pyqtSlot`.

File: guilib/measure/classes/abc.py [hunk 22]
```diff
@@ -0,0 +1,603 @@
...
+    @abstractmethod
+    def list_devices(self):
+        """List all devices of this class.
+
+        This method must return a list of SettingsInfo instances. Each
+        device is represented by a single SettingsInfo instance. The
+        SettingsInfo object must contain a .unique_name and can contain
+        .more information as a dict. The information contained within
+        a SettingsInfo must be enough to determine a suitable settings
+        file for the device from it.
```
Comment (michele-riva) [inline guilib/measure/classes/abc.py:572]: Perhaps we should be more explicit as to what "suitable" means? It is a bit generic.

File: guilib/measure/classes/abc.py [hunk 23]
```diff
@@ -0,0 +1,603 @@
...
+
+        Returns
+        -------
+        devices : list
+            Each element is a SettingsInfo instance containing the name
```
Comment (michele-riva) [inline guilib/measure/classes/abc.py:577]:
```suggestion
            Each item is a SettingsInfo instance containing the name
```

**Review by michele-riva — CHANGES_REQUESTED** [review]
Sorry Flo, some more suggestions/considerations now that I went again through the whole thing.

File: guilib/measure/camera/imagingsource.py [hunk 24]
```diff
@@ -529,6 +533,89 @@ def close(self):
...
+                                 fallback=None)
+        if match_exactly:
+            return (1,) if camera_name == obj_info.unique_name else ()
+        if camera_name is None:
+            return ()
+        camera_name_re = device_name_re(camera_name)
+        return (1,) if camera_name_re.match(obj_info.unique_name) else ()
+
+    @classmethod
+    def is_settings_for_this_class(cls, config):
```
[the remark this answers is not in the document]
Comment (FlorianDoerr) [reply guilib/measure/camera/imagingsource.py:602]: I would only generalize this part once we see we can use it for other cameras aswell. Until then I would leave it here.

[the remark this answers is not in the document]
Comment (michele-riva) [reply guilib/measure/camera/imagingsource.py:604]: Sounds good. Then perhaps add a TODO next to this and one in camera.abc (probably at the top is enough) so we can find it again if we need it.

[the remark this answers is not in the document]
Comment (FlorianDoerr) [reply guilib/measure/camera/imagingsource.py:604]: Will do

**Review by michele-riva — CHANGES_REQUESTED** [review]
Thanks for the good work @FlorianDoerr. A few more minor changes. As discussed in person, those that are likely going to give conflicts can be done directly in the soon-to-be-completed #263. Please then resolve the conversations by referring to the SHA of the commit that implemented the change (even if it is in #263).

File: guilib/measure/classes/abc.py [hunk 25]
```diff
@@ -0,0 +1,619 @@
...
+
+from viperleed.guilib.measure import hardwarebase
+from viperleed.guilib.measure.hardwarebase import DEFAULTS_PATH
+from viperleed.guilib.measure.hardwarebase import ViPErLEEDErrorEnum
+from viperleed.guilib.measure.hardwarebase import emit_error
+from viperleed.guilib.measure.classes.settings import (
+    NoDefaultSettingsError,
+    NoSettingsError,
+    SettingsError,
+    TooManyDefaultSettingsError,
```
[the remark this answers is not in the document]
Comment (michele-riva) [reply guilib/measure/classes/abc.py:42]: Actually, not needed if you do as I suggest below concerning subclassing of `DefaultSettingsError.`

**Review by michele-riva — APPROVED** [review]
Looks good!

**Pull request merged**